PM Sender

eladmallel opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Adds the PM `Sender` interface and implementation, allowing Bs to create PM tickets based on params received from O (B/O PM param exchange has yet to be implemented).

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
- `pm.Sender` with support for `StartSession` and `CreateTicket`
- `StartSession` begins a PM session with specific params, e.g. `recipientRandHash`, `winProb` and a few more
- `CreateTicket` creates the next valid ticket given a specific session
- Refactored `AccountManager` to be an interface so it can be stubbed out in `sender_test.go`

**How did you test each of these updates (required)**

- Unit tests for `defaultSender`, which is the only `Sender` implementation as of now
- Used `go test -race` and made sure the race condition detector did not find any race conditions

**Does this pull request close any open issues?**

- Fixes #629 

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

eladmallel committed:
PM: Sender WIP

Made progress towards a `Sender` that allows its users to:

* Create PM sessions
* Create PM tickets in the context of a specific session, at integrity with session params

Had to commit before finishing everything because in order to best test
the `CreateTicket` function I must refactor `AccountManager` such that
I can stub out the interface.

Once I get the refactoring done I can test for the ticket signature
more easily, as well as add some additional tests.

eladmallel committed:
ETH: AccountManager refactor to interface

So we can stub out `AccountManager` in tests. The immediate need is
for `pm.Sender`, to test for ticket signing.

eladmallel committed:
PM: Sender happy flow test with ticket sig

eladmallel committed:
Git: added livepeer binary to gitignore

Because I think it's convenient to be able to build our binary into the
main project folder.

eladmallel committed:
PM: Sender CreateTicket concurrency test

eladmallel committed:
PM: resolve race condition detector error

Thanks to Yondon's discovery of this tool:
https://blog.golang.org/race-detector

I ran the Sender tests with the race check flag `go test -race`, and
decided to fix the potential race in one of the stubs to keep things
clean all around.

eladmallel committed:
PM: Sender linter warnings fixes

yondonfu commented on pm/helpers.go:
```diff
@@ -0,0 +1,11 @@
+}
+
+func hashToHex(hash ethcommon.Hash) string {
+	return ethcommon.ToHex(hash.Bytes())
```
Looks like the `Hash` type has a [Hex()](https://github.com/ethereum/go-ethereum/blob/master/common/types.go#L72) method that converts a `Hash` type to a hex string which we can use instead of defining an additional helper here.

eladmallel replied:
Great point, fixing.

j0sh replied:
The Ethereum `Hex` and related methods have some unexpected behavior: zero-values produce "0" and it appends "0x" regardless of the value. The latter might be defensible as a question of style (note we frequently stripped the leading "0x" in most cases for the v1 protocol), but I don't believe the former is.

I've been starting to use golang's built-in `encoding/hex` instead, which should be fine for situations where we don't need compatibility with the Ethereum implementation. See the earlier discussion around this: https://github.com/livepeer/go-livepeer/pull/604/files#r239601925

eladmallel replied:
Thanks for this reference - I agree this is confusing. My followup on this is similar to your other note on this PR - I'd love to have a barebones helpers package that depends on nothing but golang and vendor packages such that all of our packages can import it, and there we can gradually refactor all these heavily reused functions, and resolve the "ToHex" implementation discussion once and for all.

wdyt?

yondonfu commented on pm/helpers.go:
```diff
@@ -0,0 +1,11 @@
+import ethcommon "github.com/ethereum/go-ethereum/common"
+
+func hexToHash(in string) ethcommon.Hash {
+	return ethcommon.BytesToHash(ethcommon.FromHex(in))
```
Looks like there is already a [HexToHash()](https://github.com/ethereum/go-ethereum/blob/master/common/types.go#L63) in `go-ethereum/common` that converts a hex string into a `Hash` type which we can use instead of defining an additional helper here.

eladmallel replied:
Great, fixing!

yondonfu commented on pm/sender.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	session := tempSession.(*session)
+
+	senderNonce := atomic.AddUint64(&session.senderNonce, 1)
```
Since `atomic.AddUint64` mutates `session.senderNonce` and returns the updated value, I think the first ticket for a sender would always have `senderNonce = 1`. Should we change this so that the first ticket for a sender always has `senderNonce = 0`?

eladmallel replied:
We could do that by assigning `senderNonce - 1` to the ticket, but I thought that would be more confusing, and didn't really see a problem with nonce starting from 1 and not 0.

Another solution could be to assign `-1` as the first value of `session.senderNonce` but that also felt clunky and confusing.

Do you have a solution in mind that is easy to read and still uses the `atomic` functionality?

yondonfu replied:
I guess my initial reaction was why wouldn't `senderNonce` start at 0 (i.e. is there some reason for it not to). However, I don't feel particularly strongly about this and I think starting a nonce at 0 or 1 is ultimately a matter of preference (as you mentioned, I don't think there are positive/negative implications for either relative to the other) so feel free to go with what you're the most comfortable with.

If you did want to start `senderNonce` at 0, could you create the ticket prior to updating `session.senderNonce`, so the ticket created always uses the current nonce value prior to its update?

eladmallel replied:
I believe the flow you have in mind could result in a race condition, since more than one goroutine could run the same lines of code in parallel that take the current value of `senderNonce`, say 0, and then when the increment it, one will increment to 1, the other to 2, because we're using atomic increment, but both tickets would have `senderNonce` of zero.

Does this make sense?

Basically my thinking is that if we use this atomic increment approach incrementing before ticket creation is the best approach, and if we used a lock then we could definitely increment after ticket creation and start with `senderNonce = 0`.

yondonfu replied:
Ah yeah I think you're right about the race condition.

I'm ok with starting `senderNonce = 1` if that means we can avoid implementing an additional lock.

yondonfu commented on pm/sender_test.go:
```diff
@@ -0,0 +1,258 @@
+		}()
+	}
+	// waiting for all session to be created
+	<-ch
```
Instead of waiting for a completion signal using a channel here and a shared counter, perhaps using [sync.WaitGroup](https://golang.org/pkg/sync/#WaitGroup) to wait for a set of goroutines to complete would be simpler and easier to read?

eladmallel replied:
Beautiful, thanks for teaching me this awesome new way.

yondonfu commented on pm/sender.go:
```diff
@@ -0,0 +1,82 @@
+
+	tempSession, ok := s.sessions.Load(sessionID)
+	if !ok {
+		return nil, nil, nil, errors.Errorf("cannot create a ticket for an unknown session: %+v", sessionID)
```
Should we use an error constant here for ease of testing?

eladmallel replied:
That could work. However, before I go for it, I'm curious as to how you would resolve the string const with the idea of adding runtime variable values to the error string?

For example, if the const string were `"cannot create a ticket for an unknown session: %+v"`, it would not really help us with testing.

An alternative I can think of is to concatenate the const error message and the variable in question, but I was having resistance to this approach because it feels unidiomatic.

wdyt?

eladmallel replied:
Now that we've discussed the idea of erring on the side of detailed errors for now - would love to get your take on if and how you would change this line of code?

yondonfu replied:
I think if we're using dynamic error strings, we'll have to use string matching while testing anyway so I think the way this error is returned seems fine now.

However, I don't think we need to use `%+v` here - IIUC that verb is only necessary when we need to log an error along with its stack trace of wrapped errors according to https://dave.cheney.net/2016/06/12/stack-traces-and-the-errors-package

eladmallel replied:
yeah, I think you're right. getting rid of the `%+v`.

yondonfu commented on pm/sender_test.go:
```diff
@@ -0,0 +1,258 @@
+
+func TestCreateTicket_GivenValidSessionId_UsesSessionParamsInTicket(t *testing.T) {
+	sender := defaultNewSender(t)
+	am := sender.accountManager.(*stubAccountManager)
```
Could we have the `stubAccountManager` instance available here and add exported methods (or export the relevant fields) for `stubAccountManager` for setting these boolean flags to avoid using a type assertion?

eladmallel replied:
Do you mean for example, instead of setting `saveSignRequest`, have a function like `ShouldSaveSignRequest(bool)`?

For the case of a stub I don't think we gain much by adding a function that simply wraps a boolean, that's why I chose to stick with this approach.

What are the benefits you see us having from adding these wrapper methods?

As for the type assertion, are you suggesting that `defaultNewSender` should return both sender and accountManager instead of just the sender? - I think that can definitely work. I avoided this at first because it kept all the other tests that don't need the accountManager a bit cleaner.
Do you think it's worth adding the extra return value?

yondonfu replied:
> For the case of a stub I don't think we gain much by adding a function that simply wraps a boolean, that's why I chose to stick with this approach.

I actually just realized that since we're writing tests internal to the package, we can just directly set the unexported fields of `stubAccountManager` so ignore my comment on using an exported wrapper function!

> As for the type assertion, are you suggesting that defaultNewSender should return both sender and accountManager instead of just the sender?

I was actually thinking that for this test you could manually instantiate `stubAccountManager` and `sender` so that you can directly use `stubAccountManager` within the test

eladmallel replied:
I prefer the cost of the type assertion for the benefit of less pain if/when we need to change the way we instantiate Sender (e.g. add more constructor params).

Does this consideration make the current approach reasonable, or do you still think it's better to manually instantiate sender and accountManager in this test function?

yondonfu replied:
Hm yeah that's a good consideration. With that in mind I think keeping the type assertion + the current approach for instantiation is reasonable for now

yondonfu commented on pm/sender_test.go:
```diff
@@ -0,0 +1,258 @@
+	ticket, actualSeed, actualSig, err := sender.CreateTicket(sessionID)
+
+	if err != nil {
+		t.Errorf("error tryint to create a ticket: %v", err)
```
"tryint" -> "trying"?

yondonfu commented on pm/sender_test.go:
```diff
@@ -0,0 +1,258 @@
+	if actualSeed != ticketParams.Seed {
+		t.Errorf("expected actual seed %d to be %d", actualSeed, ticketParams.Seed)
+	}
+	if !reflect.DeepEqual(actualSig, am.signResponse) {
```
Thoughts on using [bytes.Equal](https://golang.org/pkg/bytes/#Equal) here for clarity on types being compared?

yondonfu commented on pm/sender_test.go:
```diff
@@ -0,0 +1,258 @@
+		}()
+	}
+	// waiting for all session to be created
+	<-ch
```
Echoing previous question/comment re: using sync.WaitGroup

eladmallel committed:
PM: cleanup

Removed this helper method that was unnecessary thanks to Yondon's
feedback.

eladmallel committed:
PM: cleanup

Removed unnecessary helper thanks to Yondon's feedback.

eladmallel committed:
PM: test cleanup

eladmallel committed:
PM: test cleanup

yondonfu commented on pm/sender.go:
```diff
@@ -0,0 +1,82 @@
+	ticketParams TicketParams
+}
+
+type defaultSender struct {
```
If we're not exporting this struct right now i.e. consumers of the `pm` package wouldn't know anything about it, why not just call it `sender`?

yondonfu commented on eth/accountmanager.go:
```diff
@@ -22,14 +22,23 @@ var (
+	Account() accounts.Account
+}
+
+type DefaultAccountManager struct {
```
I see that we're using the naming convention of prepending default to the interface name for the main implementation of an interface. Why is `DefaultAccountManager` exported, but `defaultSender` in `sender.go` not exported?

eladmallel committed:
PM: cleanup

eladmallel committed:
PM: struct renaming

Thanks to PR feedback realized that `DefaultAccountManager` did not need
to be exported, and also clarified that unless we're exporting the
default struct implementation, there's no need for the `default` prefix,
so dropped that as well for both `accountManager` and `sender`.

yondonfu reviewed (approved):
LGTM 🚢

eladmallel committed:
PM: added pm folder to test script

eladmallel committed:
PM: added missing go get

j0sh commented on pm/sender_test.go:
```diff
@@ -0,0 +1,249 @@
+	return res
+}
+
+func randBytes(size int) ([]byte, error) {
```
Curious why this implementation was taken rather than the one here; this is nice because it doesn't return an error, so there's less boilerplate to handle in the above helpers.

https://github.com/livepeer/go-livepeer/blob/94c0238ddb5f73c6142d5a622d1de1a6fe1dd730/core/streamdata.go#L49-L55

eladmallel replied:
No reason except my not being aware of this implementation. Practically speaking if I really wanted to reuse this code I would have to refactor it to a super basic package that all packages can import, because in the current state if I just used it in `pm` it would result in a dependency cycle.

What do you think about creating such a package, and how would you like us to capture all these little refactoring ideas that come up here?

j0sh replied:
Ah I was referring more to the construct of generating the bytes via iterated `rand.Uint32` rather than `rand.Read` since the former doesn't throw an error. That seems like it would simplify a lot of the invocations of `randBytes`. Agree that trying to use `core.RandomIDGenerator` as-is would be complex. We actually have three or four random-bytes helpers scattered through the code 🙃 

I like the idea of a package for helpers, but it seems like `common` should mostly already be "that" package. The DB stuff is an interesting exception there, but I have some further thoughts which I'll elaborate in the [other thread](https://github.com/livepeer/go-livepeer/pull/655#discussion_r244243772).

eladmallel replied:
I'm still of the opinion that we would benefit from separating DB code out of `common`, and would still love to keep the `pm` package as decoupled as possible from any Livepeer-specific assumptions, so this implementation could potentially be reused as an open source package later.

To summarize I am taking the conversation here and in the other thread as potential refactoring TODOs between MVS and the next milestone, since I'd love to discuss it a bit further before we land on a specific action plan.
